lib/SparkPost/SparkPost.php

```diff
@@ -1,60 +1,137 @@
-	 * @throws \Exception
-	 */
-	public static function setConfig(array $configMap) {
-		//check for API key because its required	
-		if (isset($configMap['key'])){
-			$key = trim($configMap['key']);
-			if(empty($key)){
-				throw new \Exception('You must provide an API key');
-			}
-		} else {
-			throw new \Exception('You must provide an API key');
-		}
-		self::$config = self::$defaults;
-		foreach ($configMap as $configOption => $configValue) {
-			if(key_exists($configOption, self::$config)) {
-				self::$config[$configOption] = $configValue;
-			}
-		}
-	}
-	
-	/**
-	 * Retrieves the configuration that was previously setup by the user
-	 * @throws \Exception
-	 */
-	public static function getConfig() {
-		if (self::$config === null) {	
-			throw new \Exception('No configuration has been provided');
-		}
-		return self::$config;
-	}
-	
-	public static function unsetConfig() {
-		self::$config = NULL;
-	}
+
+  public $transmission;
+
+  /**
+   * @dec connection config for making requests.
+   */
+  private $config;
+
+  /**
+   * @desc Ivory\HttpAdapter\HttpAdapterInterface to make requests through.
+   */
+  public $httpAdapter;
+
+  /**
+   * @desc Default config values. Passed in values will override these.
+   */
+  private static $apiDefaults = [
+		'host'=>'api.sparkpost.com',
+		'protocol'=>'https',
+		'port'=>443,
+		'strictSSL'=>true,
+		'key'=>'',
+		'version'=>'v1'
+	];
+
+  /**
+   * @desc sets up httpAdapter and config
+   *
+   * Sets up instances of sub libraries.
+   *
+   * @param Ivory\HttpAdapter $httpAdapter - An adapter for making http requests
+   * @param Array $settingsConfig - Hashmap that contains config values for the SDK to connect to SparkPost
+   */
+	public function __construct($httpAdapter, $settingsConfig) {
+    //config needs to be setup before adapter because of default adapter settings
+    $this->setConfig($settingsConfig);
+    $this->setHttpAdapter($httpAdapter);
+
+    $this->transmission = new Transmission($this);
+  }
+
+
+
+  /**
+   * Creates an unwrapped api interface for endpoints that aren't yet supported.
+   * The new resource is attached to this object as well as returned
+   * @return SparkPost\APIResource - the unwrapped resource
+   */
+  public function setupUnwrapped (string $endpoint) {
+    $this->{$endpoint} = new APIResource($this);
+    $this->{$endpoint}->endpoint = $endpoint;
+
+    return $this->{$endpoint};
+  }
+
+  /**
+   * @desc Merges passed in headers with default headers for http requests
+   * @return Array - headers to be set on http requests
+   */
+  public function getHttpHeaders(Array $headers = null) {
+    $defaultOptions = [
+      'Authorization' => $this->config['key'],
+      'Content-Type' => 'application/json',
+    ];
+
+    // Merge passed in headers with defaults
+    if (!is_null($headers)) {
+      foreach ($headers as $header => $value) {
+        $defaultOptions[$header] = $value;
+      }
+    }
```

Comment:
So that code can't be hit because I'm only calling this in APIResource->call and I'm not passing anything into it there.  I mean somebody could call it on their own but theres no point to it.  It wouldn't do anything.